### PR TITLE
feat: scroll-controls

### LINKED ISSUE #63: ScrollControls
https://github.com/pmndrs/drei#scrollcontrols

### PR DESCRIPTION
**Pull request by JaimeTorrealba**
close #63

### CODE REVIEW

File: docs/guide/controls/scroll-controls.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
+# ScrollControls
+
```
Comment (alvarosabu) [inline docs/guide/controls/scroll-controls.md:2]: Can we embed an example here?

File: src/utils/index.ts [hunk 2]
```diff
@@ -30,3 +32,9 @@ export function hasSetter(obj: any, prop: string): boolean {
   const setterName = `set${prop[0].toUpperCase()}${prop.slice(1)}`
   return obj[setterName] !== undefined
 }
+
+export function greaterThanZero(key: string, value: number) {
```
Comment (alvarosabu) [inline src/utils/index.ts:36]: I will remove this from utils, utils should be plain typescript functions without side effects or use of composables (useLogger).

The name is also miss-leading, `greaterThanZero` implies a boolean (Is this parameter `greaterThanZero` ?) when the function actually logs a warning. 

See my comment on ScrollControls

File: src/core/controls/ScrollControls.vue [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    pages: 4,
+    distance: 4,
+    smoothScroll: 0.1,
+    horizontal: false,
+    htmlScroll: false,
+  },
+)
+
+greaterThanZero('smoothScroll', props.smoothScroll)
+greaterThanZero('pages', props.pages)
```
Comment (alvarosabu) [inline src/core/controls/ScrollControls.vue:69]:
```suggestion
if(props.smoothScroll < 0) logWarning('SmoothControl must be greater than zero')
if(props.pages < 0) logWarning('Pages must be greater than zero')
```

File: src/core/controls/ScrollControls.vue [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+const progress = ref(0)
+const progressScroll = ref(0)
+const scrollNodeY = ref(0)
+const direction = props.horizontal ? 'x' : 'y'
+
+const emit = defineEmits(['update:modelValue'])
+
+const unWatch = watch(
+  () => state.camera,
+  value => {
```
Comment (alvarosabu) [inline src/core/controls/ScrollControls.vue:93]: Add a Type to value

File: src/core/controls/ScrollControls.vue [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    if (initialized.value) {
+      unWatch()
+      return
+    }
+    initCameraPos = props.horizontal ? value.position.x || 0 : value.position.y || 0
```
Comment (alvarosabu) [inline src/core/controls/ScrollControls.vue:98]: Screenshot 2023-07-07 at 17 37 18

Try `value?.position.x` and `value?.position.y`

File: src/core/controls/ScrollControls.vue [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+onLoop(() => {
+  if (state.camera?.position) {
+    state.camera.position[direction] +=
+      (progress.value * props.distance - state.camera.position[direction] + initCameraPos) * props.smoothScroll
+    if (wrapperRef.value.children.length > 0) {
+      wrapperRef.value.position[direction] +=
+        (progress.value * props.distance - state.camera.position[direction] + initCameraPos) * props.smoothScroll
+    }
+  }
```
Comment (alvarosabu) [inline src/core/controls/ScrollControls.vue:174]:
```suggestion
if (state.camera?.position) {
    const delta =
      (progress.value * props.distance - state.camera.position[direction] + initCameraPos) * props.smoothScroll
    state.camera.position[direction] += delta
    if (wrapperRef.value.children.length > 0) {
      wrapperRef.value.position[direction] += delta
    }
  }
```

**Review by alvarosabu — CHANGES_REQUESTED** [review]
Great work chamo, just some minor things to fix and we can merge this

File: docs/guide/controls/scroll-controls.md (hunk 1 above)
Comment (JaimeTorrealba) [reply docs/guide/controls/scroll-controls.md:2]: How do you think is the best way to do it?

Currently, any abstraction in cientos have one examples (I need to create a task to add examples to all of them)

File: src/core/controls/ScrollControls.vue (hunk 4 above)
Comment (JaimeTorrealba) [reply src/core/controls/ScrollControls.vue:95]: If I try to add the type Camera, the linter throws errors

Comment (alvarosabu) [reply src/core/controls/ScrollControls.vue:95]: The linter is throwing errors as well on the pipeline

**Pull request merged**